File: app/src/main/java/com/dhpcs/liquidity/fragment/TransferToPlayerDialogFragment.java
package com.dhpcs.liquidity.fragment;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckedTextView;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.dhpcs.liquidity.boardgame.BoardGame.Identity;
import com.dhpcs.liquidity.boardgame.BoardGame.IdentityWithBalance;
import com.dhpcs.liquidity.boardgame.BoardGame.Player;
import com.dhpcs.liquidity.R;
import com.dhpcs.liquidity.activity.BoardGameActivity;
import com.dhpcs.liquidity.model.AccountId;
import com.dhpcs.liquidity.model.MemberId;

import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Currency;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import scala.Option;
import scala.collection.JavaConversions;
import scala.util.Either;

public class TransferToPlayerDialogFragment extends DialogFragment {

    public interface Listener {

        void onTransferValueEntered(Identity from, List<Player> to, BigDecimal transferValue);

    }

    public static final String TAG = "transfer_to_player_dialog_fragment";

    private static final String ARG_IDENTITIES = "identities";
    private static final String ARG_PLAYERS = "players";
    private static final String ARG_CURRENCY = "currency";
    private static final String ARG_FROM = "from";
    private static final String ARG_TO = "to";

    private static final String EXTRA_TO_LIST = "to_list";

    private static class PlayersAdapter extends ArrayAdapter<Player> {

        public PlayersAdapter(Context context, List<Player> players) {
            super(context, android.R.layout.simple_spinner_item, players);
            setDropDownViewResource(
                    android.R.layout.simple_spinner_dropdown_item
            );
        }

        private View bindView(TextView textView, Player player) {
            textView.setText(
                    BoardGameActivity.formatNullable(getContext(), player.member().name())
            );
            return textView;
        }

        @Override
        public View getDropDownView(int position, View convertView, ViewGroup parent) {
            return bindView(
                    (TextView) super.getDropDownView(position, convertView, parent),
                    getItem(position)
            );
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            return bindView(
                    (TextView) super.getView(position, convertView, parent),
                    getItem(position)
            );
        }

    }

    private static class IdentitiesAdapter extends ArrayAdapter<IdentityWithBalance> {

        public IdentitiesAdapter(Context context, List<IdentityWithBalance> identities) {
            super(context, android.R.layout.simple_spinner_item, identities);
            setDropDownViewResource(
                    android.R.layout.simple_spinner_dropdown_item
            );
        }

        private View bindView(TextView textView, IdentityWithBalance identity) {
            textView.setText(
                    identity.isBanker() ?
                            BoardGameActivity.formatNullable(
                                    getContext(),
                                    identity.member().name()
                            )
                            :
                            getContext().getString(
                                    R.string.identity_format_string,
                                    BoardGameActivity.formatNullable(
                                            getContext(),
                                            identity.member().name()
                                    ),
                                    BoardGameActivity.formatCurrencyValue(
                                            getContext(),
                                            identity.balanceWithCurrency()._2(),
                                            identity.balanceWithCurrency()._1()
                                    )
                            )
            );
            return textView;
        }

        @Override
        public View getDropDownView(int position, View convertView, ViewGroup parent) {
            return bindView(
                    (TextView) super.getDropDownView(position, convertView, parent),
                    getItem(position)
            );
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            return bindView(
                    (TextView) super.getView(position, convertView, parent),
                    getItem(position)
            );
        }

    }

    public static TransferToPlayerDialogFragment newInstance(
            scala.collection.Iterable<IdentityWithBalance> identities,
            scala.collection.Iterable<? extends Player> players,
            Option<Either<String, Currency>> currency,
            IdentityWithBalance from,
            Player to) {
        TransferToPlayerDialogFragment transferToPlayerDialogFragment =
                new TransferToPlayerDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(
                ARG_IDENTITIES,
                new ArrayList<>(
                        JavaConversions.bufferAsJavaList(
                                identities.<Identity>toBuffer()
                        )
                )
        );
        args.putSerializable(
                ARG_PLAYERS,
                new ArrayList<>(
                        JavaConversions.bufferAsJavaList(
                                players.<Player>toBuffer()
                        )
                )
        );
        args.putSerializable(ARG_CURRENCY, currency);
        args.putSerializable(ARG_FROM, from);
        args.putSerializable(ARG_TO, to);
        transferToPlayerDialogFragment.setArguments(args);
        return transferToPlayerDialogFragment;
    }

    private Listener listener;

    private Option<Either<String, Currency>> currency;
    private IdentityWithBalance from;
    private Player to;
    private ArrayList<Player> toList;

    private ArrayAdapter<IdentityWithBalance> identitiesSpinnerAdapter;
    private ArrayAdapter<Player> playersSpinnerAdapter;

    private TextView textViewValueError;
    private TextView textViewFromError;
    private Button buttonPositive;

    private BigDecimal value;

    private scala.collection.immutable.Map<MemberId, IdentityWithBalance> identities;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        listener = (Listener) activity;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        @SuppressWarnings("unchecked") List<IdentityWithBalance> identities =
                (List<IdentityWithBalance>) getArguments().getSerializable(ARG_IDENTITIES);
        @SuppressWarnings("unchecked") List<Player> players =
                (List<Player>) getArguments().getSerializable(ARG_PLAYERS);
        @SuppressWarnings("unchecked") Option<Either<String, Currency>> currency =
                (Option<Either<String, Currency>>) getArguments().getSerializable(ARG_CURRENCY);
        this.currency = currency;
        from = (IdentityWithBalance) getArguments().getSerializable(ARG_FROM);
        to = (Player) getArguments().getSerializable(ARG_TO);
        @SuppressWarnings("unchecked") ArrayList<Player> toList =
                to != null ? null : (savedInstanceState != null
                        ?
                        (ArrayList<Player>) savedInstanceState.getSerializable(EXTRA_TO_LIST)
                        :
                        new ArrayList<Player>());
        this.toList = toList;

        Comparator<Player> playerComparator = BoardGameActivity.playerComparator(getActivity());
        identitiesSpinnerAdapter = new IdentitiesAdapter(getActivity(), identities);
        identitiesSpinnerAdapter.sort(playerComparator);
        playersSpinnerAdapter = new PlayersAdapter(getActivity(), players);
        playersSpinnerAdapter.sort(playerComparator);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        @SuppressLint("InflateParams") View view = getActivity().getLayoutInflater().inflate(
                R.layout.fragment_transfer_to_player_dialog,
                null
        );

        TextView textViewCurrency = (TextView) view.findViewById(R.id.textview_currency);
        final EditText editTextValue = (EditText) view.findViewById(R.id.edittext_value);
        final TextView editTextScaledValue = (TextView)
                view.findViewById(R.id.textview_scaled_value);
        textViewValueError = (TextView) view.findViewById(R.id.textview_value_error);
        Spinner spinnerFrom = (Spinner) view.findViewById(R.id.spinner_from);
        textViewFromError = (TextView) view.findViewById(R.id.textview_from_error);
        LinearLayout linearLayoutTo = (LinearLayout) view.findViewById(R.id.linearlayout_to);
        Spinner spinnerTo = (Spinner) view.findViewById(R.id.spinner_to);

        final AlertDialog alertDialog = new AlertDialog.Builder(getActivity())
                .setTitle(
                        getString(
                                R.string.enter_transfer_details
                        )
                )
                .setView(view)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(
                        R.string.ok,
                        new DialogInterface.OnClickListener() {

                            @Override
                            public void onClick(DialogInterface dialog, int whichButton) {
                                if (listener != null) {
                                    listener.onTransferValueEntered(
                                            from,
                                            to != null ? Collections.singletonList(to) : toList,
                                            value
                                    );
                                }
                            }

                        }
                )
                .create();

        editTextValue.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @SuppressLint("SetTextI18n")
            @Override
            public void afterTextChanged(Editable s) {

                DecimalFormat numberFormat = (DecimalFormat) NumberFormat.getNumberInstance();

                try {
                    value = new BigDecimal(
                            s.toString().replace(
                                    String.valueOf(
                                            numberFormat.getDecimalFormatSymbols()
                                                    .getGroupingSeparator()
                                    ),
                                    ""
                            )
                    );
                } catch (IllegalArgumentException e) {
                    value = null;
                }

                if (value == null) {

                    editTextValue.removeTextChangedListener(this);
                    editTextValue.setText(null);
                    editTextValue.addTextChangedListener(this);
                    editTextScaledValue.setText(null);

                } else {

                    StringBuilder trailingZeros = new StringBuilder();
                    int currentDecimalSeparatorIndex = s.toString().indexOf(
                            numberFormat.getDecimalFormatSymbols().getDecimalSeparator()
                    );
                    if (currentDecimalSeparatorIndex != -1) {
                        for (int i = currentDecimalSeparatorIndex + 1; i < s.length(); i++) {
                            if (s.charAt(i)
                                    == numberFormat.getDecimalFormatSymbols().getZeroDigit()) {
                                trailingZeros.append(
                                        numberFormat.getDecimalFormatSymbols().getZeroDigit()
                                );
                            } else {
                                trailingZeros.setLength(0);
                            }
                        }
                        numberFormat.setDecimalSeparatorAlwaysShown(true);
                    }

                    int currentSelection = editTextValue.getSelectionStart();
                    int currentLength = editTextValue.getText().length();

                    numberFormat.setMaximumFractionDigits(value.scale());
                    numberFormat.setMinimumFractionDigits(0);

                    editTextValue.removeTextChangedListener(this);
                    editTextValue.setText(numberFormat.format(value) + trailingZeros);
                    editTextValue.addTextChangedListener(this);

                    int updatedLength = editTextValue.getText().length();
                    int updatedSelection = currentSelection + updatedLength - currentLength;

                    if (updatedSelection < 0 || updatedSelection > updatedLength) {
                        editTextValue.setSelection(0);
                    } else {
                        editTextValue.setSelection(updatedSelection);
                    }

                    if (value.scaleByPowerOfTen(-3).abs().compareTo(BigDecimal.ONE) < 0) {
                        editTextScaledValue.setText(null);
                    } else {
                        editTextScaledValue.setText(
                                getString(
                                        R.string.transfer_to_player_scaled_value_format_string,
                                        BoardGameActivity.formatCurrencyValue(
                                                getActivity(),
                                                currency,
                                                value
                                        )
                                )
                        );
                    }

                }

                if (buttonPositive != null) {
                    validateInput();
                }

            }

        });
        spinnerFrom.setAdapter(identitiesSpinnerAdapter);
        spinnerFrom.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> parent,
                                       View view,
                                       int position,
                                       long id) {
                from = identitiesSpinnerAdapter.getItem(position);
                if (buttonPositive != null) {
                    validateInput();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }

        });
        spinnerTo.setAdapter(playersSpinnerAdapter);
        spinnerTo.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> parent,
                                       View view,
                                       int position,
                                       long id) {
                to = playersSpinnerAdapter.getItem(position);
                if (buttonPositive != null) {
                    validateInput();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }

        });

        textViewCurrency.setText(
                BoardGameActivity.formatCurrency(
                        getActivity(),
                        currency
                )
        );

        spinnerFrom.setSelection(identitiesSpinnerAdapter.getPosition(from));

        if (to != null) {
            spinnerTo.setSelection(playersSpinnerAdapter.getPosition(to));
        } else {
            spinnerTo.setVisibility(View.GONE);
            @SuppressWarnings("unchecked") List<Player> players =
                    (List<Player>) getArguments().getSerializable(ARG_PLAYERS);
            assert players != null;
            for (final Player player : players) {
                final CheckedTextView checkedTextViewPlayer = (CheckedTextView)
                        getActivity().getLayoutInflater().inflate(
                                android.R.layout.simple_list_item_multiple_choice,
                                linearLayoutTo,
                                false
                        );
                linearLayoutTo.addView(checkedTextViewPlayer);
                checkedTextViewPlayer.setText(
                        BoardGameActivity.formatNullable(getActivity(), player.member().name())
                );
                checkedTextViewPlayer.setChecked(toList.contains(player));
                checkedTextViewPlayer.setOnClickListener(new View.OnClickListener() {

                    @Override
                    public void onClick(View v) {
                        checkedTextViewPlayer.toggle();
                        if (checkedTextViewPlayer.isChecked()) {
                            toList.add(player);
                        } else {
                            toList.remove(player);
                        }
                        validateInput();
                    }

                });
            }
        }

        alertDialog.setOnShowListener(new DialogInterface.OnShowListener() {

            @Override
            public void onShow(DialogInterface dialog) {
                buttonPositive = alertDialog.getButton(DialogInterface.BUTTON_POSITIVE);
                try {
                    value = new BigDecimal(editTextValue.getText().toString());
                } catch (IllegalArgumentException e) {
                    value = null;
                }
                validateInput();
            }

        });

        alertDialog.getWindow().setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE
        );

        return alertDialog;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    public void onIdentitiesUpdated(
            scala.collection.immutable.Map<MemberId, IdentityWithBalance> identities) {
        this.identities = identities;
        identitiesSpinnerAdapter.clear();
        identitiesSpinnerAdapter.addAll(JavaConversions.asJavaCollection(identities.values()));
        if (buttonPositive != null) {
            validateInput();
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(EXTRA_TO_LIST, toList);
    }

    private void validateInput() {
        boolean isValid = true;
        scala.math.BigDecimal currentBalance = identities == null ?
                from.balanceWithCurrency()._1() :
                identities.apply(from.member().id()).balanceWithCurrency()._1();
        if (value == null) {
            textViewValueError.setText(null);
            isValid = false;
        } else {
            BigDecimal requiredBalance = from.isBanker()
                    ?
                    null
                    :
                    value.multiply(new BigDecimal(to != null ? 1 : toList.size()));
            if (requiredBalance != null
                    && currentBalance.bigDecimal().compareTo(requiredBalance) < 0) {
                textViewValueError.setText(
                        getString(
                                R.string.transfer_value_invalid_format_string,
                                BoardGameActivity.formatCurrencyValue(
                                        getActivity(),
                                        currency,
                                        currentBalance
                                ),
                                BoardGameActivity.formatCurrencyValue(
                                        getActivity(),
                                        currency,
                                        requiredBalance
                                )
                        )
                );
                isValid = false;
            } else {
                textViewValueError.setText(null);
            }
        }
        Set<AccountId> toAccountIds;
        if (to != null) {
            toAccountIds = Collections.singleton(to.account().id());
        } else {
            toAccountIds = new HashSet<>(toList.size());
            for (Player to : toList) {
                toAccountIds.add(to.account().id());
            }
        }
        if (toAccountIds.contains(from.account().id())) {
            textViewFromError.setText(
                    getString(
                            R.string.transfer_from_invalid_format_string,
                            BoardGameActivity.formatNullable(getActivity(), from.member().name())
                    )
            );
            isValid = false;
        } else {
            textViewFromError.setText(null);
        }
        buttonPositive.setEnabled(isValid);
    }

}

File: app/src/main/java/com/dhpcs/liquidity/fragment/TransferToPlayerDialogFragment.kt
package com.dhpcs.liquidity.fragment

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.dhpcs.liquidity.BoardGame
import com.dhpcs.liquidity.LiquidityApplication
import com.dhpcs.liquidity.R
import com.dhpcs.liquidity.activity.MainActivity
import com.dhpcs.liquidity.activity.MainActivity.Companion.maybeLiveData
import com.dhpcs.liquidity.activity.MainActivity.Companion.observableLiveData
import io.reactivex.subjects.BehaviorSubject
import java.math.BigDecimal
import java.text.DecimalFormat
import java.text.NumberFormat

class TransferToPlayerDialogFragment : AppCompatDialogFragment() {

    companion object {

        const val TAG = "transfer_to_player_dialog_fragment"

        private const val ARG_FROM_IDENTITY_ID = "from_identity_id"
        private const val ARG_TO_PLAYER_ID = "to_player_id"

        fun newInstance(
                from: BoardGame.Companion.Identity,
                to: BoardGame.Companion.Player?): TransferToPlayerDialogFragment {
            val transferToPlayerDialogFragment = TransferToPlayerDialogFragment()
            val args = Bundle()
            args.putString(ARG_FROM_IDENTITY_ID, from.memberId)
            args.putString(ARG_TO_PLAYER_ID, to?.memberId)
            transferToPlayerDialogFragment.arguments = args
            return transferToPlayerDialogFragment
        }

        private class PlayersAdapter
        internal constructor(context: Context,
                             players: List<BoardGame.Companion.Player>
        ) : ArrayAdapter<BoardGame.Companion.Player>(
                context,
                android.R.layout.simple_spinner_item,
                players
        ) {

            init {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return bindView(
                        super.getView(position, convertView, parent) as TextView,
                        getItem(position)
                )
            }

            override fun getDropDownView(position: Int,
                                         convertView: View?,
                                         parent: ViewGroup
            ): View {
                return bindView(
                        super.getDropDownView(position, convertView, parent) as TextView,
                        getItem(position)
                )
            }

            private fun bindView(textView: TextView, player: BoardGame.Companion.Player?): View {
                textView.text = LiquidityApplication.formatNullable(context, player!!.name)
                return textView
            }

        }

        private class IdentitiesAdapter
        internal constructor(context: Context,
                             identities: List<BoardGame.Companion.Identity>
        ) : ArrayAdapter<BoardGame.Companion.Identity>(
                context,
                android.R.layout.simple_spinner_item,
                identities
        ) {

            init {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return bindView(
                        super.getView(position, convertView, parent) as TextView,
                        getItem(position)
                )
            }

            override fun getDropDownView(position: Int,
                                         convertView: View?,
                                         parent: ViewGroup
            ): View {
                return bindView(
                        super.getDropDownView(position, convertView, parent) as TextView,
                        getItem(position)
                )
            }

            private fun bindView(textView: TextView,
                                 identity: BoardGame.Companion.Identity?
            ): View {
                textView.text = if (identity!!.isBanker) {
                    LiquidityApplication.formatNullable(context, identity.name)
                } else {
                    context.getString(
                            R.string.identity_format_string,
                            LiquidityApplication.formatNullable(context, identity.name),
                            LiquidityApplication.formatCurrencyValue(
                                    context,
                                    identity.currency,
                                    identity.balance
                            )
                    )
                }
                return textView
            }

        }

    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        @SuppressLint("InflateParams") val view = requireActivity().layoutInflater.inflate(
                R.layout.fragment_transfer_to_player_dialog, null
        )

        val textViewCurrency = view.findViewById<TextView>(R.id.textview_currency)
        val editTextValue = view.findViewById<EditText>(R.id.edittext_value)
        val editTextScaledValue = view.findViewById<TextView>(R.id.textview_scaled_value)
        val textViewValueError = view.findViewById<TextView>(R.id.textview_value_error)
        val spinnerFrom = view.findViewById<Spinner>(R.id.spinner_from)
        val textViewFromError = view.findViewById<TextView>(R.id.textview_from_error)
        val linearLayoutTo = view.findViewById<LinearLayout>(R.id.linearlayout_to)
        val spinnerTo = view.findViewById<Spinner>(R.id.spinner_to)

        var value = BigDecimal.ZERO
        lateinit var buttonPositive: Button

        val fromIdentityId = arguments!!.getString(ARG_FROM_IDENTITY_ID)!!
        val fromIdentityIdSubject = BehaviorSubject.createDefault(
                fromIdentityId
        )
        val toPlayerId = arguments!!.getString(ARG_TO_PLAYER_ID)
        val toPlayerIdsSubject = BehaviorSubject.createDefault(
                if (toPlayerId != null) listOf(toPlayerId) else emptyList()
        )
        val fromSubject = BehaviorSubject.create<BoardGame.Companion.Identity>()
        val toSubject = BehaviorSubject.create<Collection<BoardGame.Companion.Player>>()

        val model = ViewModelProviders.of(requireActivity())
                .get(MainActivity.Companion.BoardGameModel::class.java)
        val alertDialog = AlertDialog.Builder(requireContext())
                .setTitle(getString(
                        R.string.enter_transfer_details
                ))
                .setView(view)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok) { _, _ ->
                    val from = fromSubject.value
                    if (from != null) {
                        toSubject.value?.forEach { to ->
                            val error = getString(
                                    R.string.transfer_to_player_error_format_string, to.name
                            )
                            model.execCommand(
                                    model.boardGame.transferToPlayer(from, to, value!!)
                            ) { error }
                        }
                    }
                }
                .create()

        fun validateInput() {
            val currentBalance = fromSubject.value?.balance
            val isValueValid = if (value == null) {
                textViewValueError.text = null
                false
            } else {
                val requiredBalance = if (fromSubject.value?.isBanker == true) {
                    null
                } else {
                    value!!.multiply(BigDecimal(toSubject.value?.size ?: 0))
                }
                if (currentBalance != null && requiredBalance != null &&
                        currentBalance < requiredBalance) {
                    model.boardGame.maybeLiveData {
                        model.boardGame.currencyObservable.firstElement()
                    }.observe(this, Observer { currency ->
                        textViewValueError.text = getString(
                                R.string.transfer_value_invalid_format_string,
                                LiquidityApplication.formatCurrencyValue(
                                        requireContext(),
                                        currency,
                                        currentBalance
                                ),
                                LiquidityApplication.formatCurrencyValue(
                                        requireContext(),
                                        currency,
                                        requiredBalance
                                )
                        )
                    })
                    false
                } else {
                    textViewValueError.text = null
                    true
                }
            }
            val toAccountIds = toSubject.value?.map { it.accountId } ?: emptyList()
            val isFromValid = if (toAccountIds.contains(fromSubject.value?.accountId)) {
                textViewFromError.text = getString(
                        R.string.transfer_from_invalid_format_string,
                        LiquidityApplication.formatNullable(
                                requireContext(),
                                fromSubject.value?.name
                        )
                )
                false
            } else {
                textViewFromError.text = null
                true
            }
            buttonPositive.isEnabled = isValueValid && isFromValid
        }

        editTextValue.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            @SuppressLint("SetTextI18n")
            override fun afterTextChanged(s: Editable) {
                val numberFormat = NumberFormat.getNumberInstance() as DecimalFormat
                value = if (s.isEmpty()) {
                    BigDecimal.ZERO
                } else {
                    BigDecimal(
                            s.toString().replace(
                                    numberFormat.decimalFormatSymbols.groupingSeparator.toString(),
                                    ""
                            )
                    )
                }
                if (value == null) {
                    editTextValue.removeTextChangedListener(this)
                    editTextValue.text = null
                    editTextValue.addTextChangedListener(this)
                    editTextScaledValue.text = null
                } else {
                    val trailingZeros = StringBuilder()
                    val currentDecimalSeparatorIndex = s.toString().indexOf(
                            numberFormat.decimalFormatSymbols.decimalSeparator
                    )
                    if (currentDecimalSeparatorIndex != -1) {
                        for (i in currentDecimalSeparatorIndex + 1 until s.length) {
                            if (s[i] == numberFormat.decimalFormatSymbols.zeroDigit) {
                                trailingZeros.append(
                                        numberFormat.decimalFormatSymbols.zeroDigit
                                )
                            } else {
                                trailingZeros.setLength(0)
                            }
                        }
                        numberFormat.isDecimalSeparatorAlwaysShown = true
                    }

                    val currentSelection = editTextValue.selectionStart
                    val currentLength = editTextValue.text.length

                    numberFormat.maximumFractionDigits = value!!.scale()
                    numberFormat.minimumFractionDigits = 0

                    editTextValue.removeTextChangedListener(this)
                    editTextValue.setText(numberFormat.format(value) + trailingZeros)
                    editTextValue.addTextChangedListener(this)

                    val updatedLength = editTextValue.text.length
                    val updatedSelection = currentSelection + updatedLength - currentLength
                    if (updatedSelection < 0 || updatedSelection > updatedLength) {
                        editTextValue.setSelection(0)
                    } else {
                        editTextValue.setSelection(updatedSelection)
                    }

                    if (value!!.scaleByPowerOfTen(-3).abs() < BigDecimal.ONE) {
                        editTextScaledValue.text = null
                    } else {
                        model.boardGame.maybeLiveData {
                            model.boardGame.currencyObservable.firstElement()
                        }.observe(this@TransferToPlayerDialogFragment, Observer { currency ->
                            editTextScaledValue.text = getString(
                                    R.string.transfer_to_player_scaled_value_format_string,
                                    LiquidityApplication.formatCurrencyValue(
                                            requireContext(),
                                            currency,
                                            value!!
                                    )
                            )
                        })
                    }
                }
                validateInput()
            }

        })

        model.boardGame.observableLiveData {
            model.boardGame.currencyObservable
        }.observe(this, Observer { currency ->
            textViewCurrency.text = LiquidityApplication.formatCurrency(requireContext(), currency)
        })

        model.boardGame.observableLiveData {
            it.identitiesObservable
        }.observe(this, Observer {
            val identitiesSpinnerAdapter = IdentitiesAdapter(
                    requireContext(),
                    it.values.sortedWith(LiquidityApplication.identityComparator(requireContext()))
            )
            spinnerFrom.adapter = identitiesSpinnerAdapter
            spinnerFrom.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(parent: AdapterView<*>,
                                            view: View?,
                                            position: Int,
                                            id: Long) {
                    fromIdentityIdSubject.onNext(
                            identitiesSpinnerAdapter.getItem(position)!!.memberId
                    )
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}

            }

            spinnerFrom.setSelection(
                    identitiesSpinnerAdapter.getPosition(
                            it[fromIdentityIdSubject.value!!]
                    )
            )
        })

        model.boardGame.observableLiveData {
            it.playersObservable
        }.observe(this, Observer {
            val playersSpinnerAdapter = PlayersAdapter(
                    requireContext(),
                    it.values.sortedWith(LiquidityApplication.playerComparator(requireContext()))
            )
            spinnerTo.adapter = playersSpinnerAdapter
            spinnerTo.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(parent: AdapterView<*>,
                                            view: View?,
                                            position: Int,
                                            id: Long) {
                    toPlayerIdsSubject.onNext(
                            listOf(playersSpinnerAdapter.getItem(position)!!.memberId)
                    )
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}

            }

            if (toPlayerId != null) {
                spinnerTo.setSelection(playersSpinnerAdapter.getPosition(it[toPlayerId]))
            } else {
                spinnerTo.visibility = View.GONE
                for (player in it.values) {
                    val checkedTextViewPlayer = requireActivity().layoutInflater.inflate(
                            android.R.layout.simple_list_item_multiple_choice,
                            linearLayoutTo,
                            false
                    ) as CheckedTextView
                    linearLayoutTo.addView(checkedTextViewPlayer)
                    checkedTextViewPlayer.text = LiquidityApplication.formatNullable(
                            requireContext(),
                            player.name
                    )
                    checkedTextViewPlayer.isChecked = toPlayerIdsSubject.value!!.contains(
                            player.memberId
                    )
                    checkedTextViewPlayer.setOnClickListener {
                        checkedTextViewPlayer.toggle()
                        val toIdentities = toPlayerIdsSubject.value ?: emptyList()
                        toPlayerIdsSubject.onNext(
                                if (checkedTextViewPlayer.isChecked) {
                                    toIdentities.plus(player.memberId)
                                } else {
                                    toIdentities.minus(player.memberId)
                                }
                        )
                    }
                }
            }
        })

        model.boardGame.observableLiveData { boardGame ->
            fromIdentityIdSubject.flatMap {
                boardGame.identitiesObservable.filter { identities ->
                    identities.values.any { identity -> identity.memberId == it }
                }.map { identities -> identities.getValue(it) }
            }
        }.observe(this, Observer {
            fromSubject.onNext(it)
            validateInput()
        })
        model.boardGame.observableLiveData { boardGame ->
            toPlayerIdsSubject.flatMap {
                boardGame.playersObservable.map { players ->
                    players.values.filter { player -> it.contains(player.memberId) }
                }
            }
        }.observe(this, Observer {
            toSubject.onNext(it)
            validateInput()
        })

        alertDialog.setOnShowListener {
            buttonPositive = alertDialog.getButton(DialogInterface.BUTTON_POSITIVE)
        }

        val window = alertDialog.window!!
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)

        return alertDialog
    }

}
